File: frontend/src/services/facturas.ts
import api from './api'

// ============ TYPES ============

export type EstadoFactura = 'pendiente' | 'pagada'

export interface Factura {
  id: string
  cliente_id: string
  proyecto_id: string
  descripcion?: string | null
  fecha_inscripcion: string
  monto: number
  fecha_pago?: string | null
  estado: EstadoFactura
  transaccion_id?: string | null
  creado_por_id: string
  observaciones?: string | null
  cliente_nombre?: string | null
  proyecto_nombre?: string | null
}

export interface FacturaCreate {
  cliente_id: string
  proyecto_id: string
  descripcion?: string
  fecha_inscripcion: string
  monto: number
  observaciones?: string
}

export interface FacturaUpdate {
  cliente_id?: string
  proyecto_id?: string
  descripcion?: string
  fecha_inscripcion?: string
  monto?: number
  observaciones?: string
}

export interface MarcarPagadaRequest {
  fecha_pago: string
  observaciones?: string
}

export interface EmpresaFacturasResumen {
  id: string
  nombre: string
  cuit?: string | null
  email?: string | null
  telefono?: string | null
  cantidad_facturas: number
  cantidad_pendientes: number
  total_pendiente: number
  total_pagado: number
}

export interface EmpresaQuickCreate {
  razon_social: string
  nombre_fantasia?: string
  cuit?: string
  email?: string
  telefono?: string
}

export interface FacturaMesItem {
  anio: number
  mes: number
  label: string
  total: number
  cantidad: number
  facturas: Factura[]
}

// ============ EMPRESAS ============

const BASE = '/facturas-a-cobrar'

export const facturasService = {
  // Empresas
  listarEmpresas: async (): Promise<EmpresaFacturasResumen[]> => {
    const { data } = await api.get(`${BASE}/empresas`)
    return data
  },

  crearEmpresa: async (payload: EmpresaQuickCreate): Promise<EmpresaFacturasResumen> => {
    const { data } = await api.post(`${BASE}/empresas`, payload)
    return data
  },

  eliminarEmpresa: async (clienteId: string): Promise<void> => {
    await api.delete(`${BASE}/empresas/${clienteId}`)
  },

  // Facturas
  listar: async (params?: {
    cliente_id?: string
    proyecto_id?: string
    estado?: EstadoFactura
    fecha_desde?: string
    fecha_hasta?: string
  }): Promise<Factura[]> => {
    const { data } = await api.get(BASE, { params })
    return data
  },

  listarPendientes: async (): Promise<Factura[]> => {
    const { data } = await api.get(`${BASE}/pendientes`)
    return data
  },

  porMes: async (params?: { cliente_id?: string; anio?: number }): Promise<FacturaMesItem[]> => {
    const { data } = await api.get(`${BASE}/por-mes`, { params })
    return data
  },

  obtener: async (id: string): Promise<Factura> => {
    const { data } = await api.get(`${BASE}/${id}`)
    return data
  },

  crear: async (payload: FacturaCreate): Promise<Factura> => {
    const { data } = await api.post(BASE, payload)
    return data
  },

  actualizar: async (id: string, payload: FacturaUpdate): Promise<Factura> => {
    const { data } = await api.put(`${BASE}/${id}`, payload)
    return data
  },

  eliminar: async (id: string): Promise<void> => {
    await api.delete(`${BASE}/${id}`)
  },

  marcarPagada: async (id: string, payload: MarcarPagadaRequest): Promise<Factura> => {
    const { data } = await api.post(`${BASE}/${id}/marcar-pagada`, payload)
    return data
  },

  marcarPendiente: async (id: string): Promise<Factura> => {
    const { data } = await api.post(`${BASE}/${id}/marcar-pendiente`)
    return data
  },
}

File: backend/app/api/v1/endpoints/facturas.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from uuid import UUID
from typing import Optional, List
from datetime import date

from app.core.deps import get_db, require_admin_or_supervisor
from app.models.usuario import Usuario
from app.schemas.factura import (
    FacturaCreate, FacturaUpdate, FacturaResponse,
    MarcarPagadaRequest,
    EmpresaFacturasResumen, EmpresaQuickCreate,
    FacturaMesItem,
)
from app.services import factura_service
from app.services.factura_service import _serializar_factura

router = APIRouter()


# ============ EMPRESAS ============

@router.get("/empresas", response_model=List[EmpresaFacturasResumen])
def listar_empresas(
    db: Session = Depends(get_db),
    _: Usuario = Depends(require_admin_or_supervisor),
):
    """Lista las empresas (clientes tipo=empresa) con resumen de sus facturas."""
    return factura_service.listar_empresas_con_resumen(db)


@router.post("/empresas", response_model=EmpresaFacturasResumen)
def crear_empresa(
    data: EmpresaQuickCreate,
    db: Session = Depends(get_db),
    _: Usuario = Depends(require_admin_or_supervisor),
):
    """Alta rapida de empresa para el modulo de facturas."""
    cliente = factura_service.crear_empresa_rapida(db, data)
    return EmpresaFacturasResumen(
        id=cliente.id,
        nombre=cliente.nombre_fantasia or cliente.razon_social,
        cuit=cliente.cuit,
        email=cliente.email,
        telefono=cliente.telefono,
        cantidad_facturas=0,
        cantidad_pendientes=0,
        total_pendiente=0.0,
        total_pagado=0.0,
    )


@router.delete("/empresas/{cliente_id}")
def eliminar_empresa(
    cliente_id: UUID,
    db: Session = Depends(get_db),
    _: Usuario = Depends(require_admin_or_supervisor),
):
    if not factura_service.eliminar_empresa(db, cliente_id):
        raise HTTPException(status_code=404, detail="Empresa no encontrada")
    return {"message": "Empresa eliminada"}


# ============ FACTURAS ============

@router.get("", response_model=List[FacturaResponse])
def listar_facturas(
    cliente_id: Optional[UUID] = None,
    proyecto_id: Optional[UUID] = None,
    estado: Optional[str] = None,
    fecha_desde: Optional[date] = None,
    fecha_hasta: Optional[date] = None,
    db: Session = Depends(get_db),
    _: Usuario = Depends(require_admin_or_supervisor),
):
    facturas = factura_service.listar_facturas(
        db, cliente_id, proyecto_id, estado, fecha_desde, fecha_hasta
    )
    return [_serializar_factura(f) for f in facturas]


@router.get("/pendientes", response_model=List[FacturaResponse])
def listar_pendientes(
    db: Session = Depends(get_db),
    _: Usuario = Depends(require_admin_or_supervisor),
):
    """Todas las facturas pendientes de todas las empresas."""
    facturas = factura_service.listar_pendientes(db)
    return [_serializar_factura(f) for f in facturas]


@router.get("/por-mes", response_model=List[FacturaMesItem])
def facturas_por_mes(
    cliente_id: Optional[UUID] = None,
    anio: Optional[int] = None,
    db: Session = Depends(get_db),
    _: Usuario = Depends(require_admin_or_supervisor),
):
    """Agrupa facturas por mes de pago (o inscripcion si pendiente)."""
    return factura_service.agrupar_por_mes(db, cliente_id, anio)


@router.post("", response_model=FacturaResponse)
def crear_factura(
    data: FacturaCreate,
    db: Session = Depends(get_db),
    usuario: Usuario = Depends(require_admin_or_supervisor),
):
    factura = factura_service.crear_factura(db, data, usuario.id)
    return _serializar_factura(factura)


@router.get("/{factura_id}", response_model=FacturaResponse)
def obtener_factura(
    factura_id: UUID,
    db: Session = Depends(get_db),
    _: Usuario = Depends(require_admin_or_supervisor),
):
    factura = factura_service.obtener_factura(db, factura_id)
    if not factura:
        raise HTTPException(status_code=404, detail="Factura no encontrada")
    return _serializar_factura(factura)


@router.put("/{factura_id}", response_model=FacturaResponse)
def actualizar_factura(
    factura_id: UUID,
    data: FacturaUpdate,
    db: Session = Depends(get_db),
    _: Usuario = Depends(require_admin_or_supervisor),
):
    factura = factura_service.actualizar_factura(db, factura_id, data)
    if not factura:
        raise HTTPException(status_code=404, detail="Factura no encontrada")
    return _serializar_factura(factura)


@router.delete("/{factura_id}")
def eliminar_factura(
    factura_id: UUID,
    db: Session = Depends(get_db),
    _: Usuario = Depends(require_admin_or_supervisor),
):
    if not factura_service.eliminar_factura(db, factura_id):
        raise HTTPException(status_code=404, detail="Factura no encontrada")
    return {"message": "Factura eliminada"}


@router.post("/{factura_id}/marcar-pagada", response_model=FacturaResponse)
def marcar_pagada(
    factura_id: UUID,
    data: MarcarPagadaRequest,
    db: Session = Depends(get_db),
    usuario: Usuario = Depends(require_admin_or_supervisor),
):
    """Marca la factura como pagada y genera la Transaccion INGRESO."""
    factura = factura_service.marcar_pagada(db, factura_id, data, usuario.id)
    if not factura:
        raise HTTPException(status_code=404, detail="Factura no encontrada")
    return _serializar_factura(factura)


@router.post("/{factura_id}/marcar-pendiente", response_model=FacturaResponse)
def marcar_pendiente(
    factura_id: UUID,
    db: Session = Depends(get_db),
    _: Usuario = Depends(require_admin_or_supervisor),
):
    """Revierte una factura pagada a pendiente y anula la transaccion asociada."""
    factura = factura_service.marcar_pendiente(db, factura_id)
    if not factura:
        raise HTTPException(status_code=404, detail="Factura no encontrada")
    return _serializar_factura(factura)
